R/cube-result.R

```diff
@@ -388,6 +411,122 @@ setMethod("margin.table", "CrunchCube", function (x, margin=NULL) {
     cubeMarginTable(x, margin)
 })
 
+#' @rdname cube-computing
+#' @export
+setMethod("collapse.dimensions", "CrunchCube", function (x, margin=NULL) {
+    # ensure that we have sensible margin input
+    selecteds <- is.selectedDimension(x@dims)
+    check_margins(margin, selecteds)
+    
+    # translate from user-cube margins to real-cube margins and establish the
+    # two groups of margins: those to collapse and those to keep
+    margins_to_collapse <- user2real(margin, cube = x)
+    margins_to_keep <- setdiff(seq_along(x@dims), margins_to_collapse)
+    
+    # if there are any _items in the margin to collapse, be wary!
+    collapsed_items <- endsWith(getDimTypes(x)[margins_to_collapse], "_items")
+
+    # grab the old cube structure and over-write it with new subset data.
+    out <- x
+    
+    # iterate through measures, collapsing the dimension(s) specified.
+    out@arrays[] <- lapply(out@arrays, function(arr){
+        # off_margins here are the margins that we want to keep, however because
+        # we treat items dimensions differently, we need to add them to the
+        # off_margins, and deal with them after we collapse by summing
+        off_margins <- margins_to_keep
+        if (any(collapsed_items)) {
```

Comment:
I don't love having two conditional branches here and at 455, is it possible to combine the two blocks of conditional code by repeating the `array <- as.array(apply(arr, off_margins, sum))` call?  Personally I find repetative code with fewer branches a bit easier to follow.

Reply:
Turns out, you don't need this first if since `collapsed_items` is a logical, so I've removed it.